protocols/floodsub/src/protocol.rs

```diff
@@ -135,13 +149,22 @@ where
     type Error = io::Error;
     type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn upgrade_outbound(self, mut socket: TSocket, _: Self::Info) -> Self::Future {
+    fn upgrade_outbound(self, socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
             let bytes = self.into_bytes();
 
-            upgrade::write_length_prefixed(&mut socket, bytes).await?;
-            socket.close().await?;
-
+            let mut framed = Framed::new(
+                socket,
+                prost_codec::Codec::<std::vec::Vec<u8>>::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            framed
+                .send(bytes)
+                .await
+                .map_err(|e| io::Error::new(std::io::ErrorKind::ConnectionRefused, e))?;
```

Comment:
What type is `e` here?

Reply:
Great question, and I'm not 100% sure. Its some kind of socket error, and I was wanting to use as general of a an error where as I could get away with. But I couldn't figure out how to get the compiler to accept a more "general" thing. VSCode kept offering suggestions of ConnectionRefused and other similar/likely errors (peer disconnected, etc.). I picked ConnectionRefused (and ConnectionReset below) knowing that we might be talking about them. ;)

I did try refactoring the `FloodsubDecodeError` to see if that could be expanded into more of a general "FloodsubError", but it really started to feel like scope creep on this ticket, at least in my opinion. I'd really like to do that approach (just feels more architecturally "correct" IMO), but I also want to keep this PR focused on the problem at hand.

I'm open to suggestions on all of this. (I also want to put some tests here, but first I need to crawl, then walk, then run :laughing: )

Reply:
I'd suggest we:

1. Create a type called `EncodeError` analogous to `DecodeError`.
2. Set `type Error = EncodeError` on line 149.
3. Remove these `map_err` calls.